### ISSUE TITLE
ERROR `context canceled` when syncing large number of Bitbucket projectKeys

### ISSUE THREAD

**Issue by mike-r-mclaughlin** [sourcegraph/sourcegraph#40153]
- **Sourcegraph version:** 3.42.0

#### Steps to reproduce:

1. Create a new Bitbucket external code host configuration with 2500 project keys
```
{
  "url": "http://bitbucketserver/",
  "token": "REDACTED",
  "username": "username",
  "projectKeys": ["aa", "ab", "ab","ac", "ad","ae","af", "ag", "ah", "ai", "aj", "ak", "al", "am", "an", "ao", "ap", "aq" ...],
}
```
2. Save and watch the repo-updater logs

#### Expected behavior:

No errors or warnings occur in the UI or in the logs

#### Actual behavior:

The UI shows a warning in the repository status (both in the cloud icon in the upper right and on the repository status page).  

Errors:
```
│ {"SeverityText":"ERROR","Timestamp":1660039706810815095,"InstrumentationScope":"service.ObservedSource","Caller":"repos/observability.go:99","Function":"github.com/sourcegraph/sourcegraph/internal/repos.(*observedSource).ListRepos.func1","Body":"source.list-repos","Resource":{"service.name":"repo-updater","service. │
│ version":"3.42.0","service.instance.id":"repo-updater-767459777c-4wwn9"},"Attributes":{"error":"17 errors occurred:\n\t* bitbucketserver.projectKeys: query=\"aq\": context canceled\n\t* bitbucketserver.projectKeys: query=\"ad\": context canceled\n\t* bitbucketserver.projectKeys: query=\"aj\": context canceled\n\t*  │
│ bitbucketserver.projectKeys: query=\"am\": context canceled\n\t* bitbucketserver.projectKeys: query=\"ae\": context canceled\n\t* bitbucketserver.projectKeys: query=\"ah\": context canceled\n\t* bitbucketserver.projectKeys: query=\"an\": context canceled\n\t* bitbucketserver.projectKeys: query=\"ap\": context cance │
│ led\n\t* bitbucketserver.projectKeys: query=\"ai\": context canceled\n\t* bitbucketserver.projectKeys: query=\"ao\": context canceled\n\t* bitbucketserver.projectKeys: query=\"ac\": context canceled\n\t* bitbucketserver.projectKeys: query=\"ak\": context canceled\n\t* bitbucketserver.projectKeys: query=\"al\": cont │
│ ext canceled\n\t* bitbucketserver.projectKeys: query=\"ag\": context canceled\n\t* bitbucketserver.projectKeys: query=\"ab\": context canceled\n\t* bitbucketserver.projectKeys: query=\"ab\": context canceled\n\t* bitbucketserver.projectKeys: query=\"aa\": context canceled"}}
```

#### Additional Details
* Internal [Slack Discussion](https://sourcegraph.slack.com/archives/C03MV6CKU1X/p1659977003016689)
* Reported by: https://github.com/sourcegraph/accounts/issues/3

/cc @jplahn @ryphil

**Comment by mike-r-mclaughlin** [comment]
Confirmation that this is still happening (per meeting with @indradhanush and @stefanhengl ):

> I added new projects to the existing config.
> Just to confirm the new projects that I added got cloned almost immediately.
> query=\"XYXXYY\": has the project key
 
 ```
{"SeverityText":"ERROR","Timestamp":1660302011958108195,"InstrumentationScope":"service.ObservedSource","Caller":"repos/observability.go:99","Function":"[github.com/sourcegraph/sourcegraph/internal/repos](http://github.com/sourcegraph/sourcegraph/internal/repos).(*observedSource).ListRepos.func1","Body":"source.list-repos","Resource":{"[service.name](http://service.name/)":"repo-updater","service.version":"3.42.0","[service.instance.id](http://service.instance.id/)":"repo-updater-b9475f9bf-2rr2s"},"Attributes":{"error":"6 errors occurred:\n\t* bitbucketserver.projectKeys: query=\"HHHH\": context canceled\n\t* bitbucketserver.projectKeys: query=\"YYYYYYY\": context canceled\n\t* bitbucketserver.projectKeys: query=\"PROJECT_NAME\": context canceled\n\t* bitbucketserver.projectKeys: query=\"XYXXYY\": context canceled\n\t* bitbucketserver.projectKeys: query=\"EMRXXX\": context canceled\n\t* bitbucketserver.projectKeys: query=\"IIIII\": context canceled"}}
{"SeverityText":"WARN","Timestamp":1660302020592399427,"InstrumentationScope":"service.PermsSyncer","Caller":"authz/perms_syncer.go:159","Function":"[github.com/sourcegraph/sourcegraph/enterprise/cmd/repo-updater/internal/authz](http://github.com/sourcegraph/sourcegraph/enterprise/cmd/repo-updater/internal/authz).(*PermsSyncer).ScheduleRepos","Body":"ScheduleRepos.disabled","Resource":{"[service.name](http://service.name/)":"repo-updater","service.version":"3.42.0","[service.instance.id](http://service.instance.id/)":"repo-updater-b9475f9bf-2rr2s"},"Attributes":{"len(repoIDs)":1}}
```

**Comment by mike-r-mclaughlin** [comment]
updating the priority from p1 to p2 since the workaround of waiting for it to fix itself is sufficient for the short term. we should definitely fix this, but we have a workaround for now.

**Comment by indradhanush** [comment]
## Update 2022-08-15

**TL;DR**: Yet to discover the root cause but I have developed some understanding behind what's causing this and how to reproduce this reliably and what should be our next steps.


- If you add enough (> 300) project keys to the code host configuration and save the config, the sync triggered will eventually start choking up the Bitbucket server and we start seeing the `context canceled` error from `repo-updater`

- At this point, reducing the number of project keys to anything but `1` continues to reproduce the error, that is if there are 3 project keys, 2 context canceled errors are seen in the logs, and if there are 2 project keys, 1 context canceled error is seen in the log

- However if only 1 project key is specified, no context canceled error is seen at all


- At this point, if you go take a short walk or make yourself a cup of coffee, 5 - 10 minutes would have passed and if you add 2 project keys, the context canceled error is no longer seen, which is what leads me to believe that once the bitbucket server is choked up, it is unable to respond to more than 1 project key at a time (this is just a thought)


**Some things that I still don't understand**

- While we see the errors in the logs, the sync job is still completed as successful, as verified by running the following query on the replica instance:

```
select * from external_service_sync_jobs where external_service_id = 2 order by started_at desc
```

- What is the root cause of the contexts canceled error? This, IMO should be our next step. That is maybe we can add some debug code in our `internal/httpcli` probably because the error seems to be coming down from the network requests


- Is one of the parent context sending a cancellation? This requires more troubleshooting


cc: @stefanhengl

**Comment by varsanojidan** [comment]
So I have a bit of a theory as to why this could be happening:
1) When listing projectKeys, we start a new goroutine for each projectKey [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/repos/bitbucketserver.go?L276&subtree=true)
2) If any of the http requests for the projectKey repos inside the goroutines error out for ANY reason, we return an error on the channel [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/repos/bitbucketserver.go?L297&subtree=true)
3) If the error is considered fatal (which includes unauthorized errors), we break the repo syncing process [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/repos/syncer.go?L606)
4) This break ends SyncExternalService function, and calls a context cancel [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/repos/syncer.go?L535)
5) The remaining inflight goroutines would then get the context canceled error (one per remaining projectKey like we see in the logs)
6) This prevents the full repo sync, so it never deletes the excluded repos

**Comment by varsanojidan** [comment]
Root cause of the context cancelled issues:

TL;DR: Its harmless.

How this happens:
All of the context cancelled errors are coming from manually triggered syncs of the codehost config:

1) Triggering a manual sync calls the `handleExternalServiceSync` function [underneath](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/repo-updater/repoupdater/server.go?L185&subtree=true)
2) `handleExternalServiceSync` calls `externalServiceValidate` [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/repo-updater/repoupdater/server.go?L272&subtree=true)
a) this function is the root cause of the context cancels
b) as part of the validation it calls `ListRepos` and immediately exits after the first result because after 1 result it considers the config "valid"
c) this triggers the cancel func [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/repo-updater/repoupdater/server.go?L286&subtree=true) which cancels all remaining inflight requests to BitBucket from `ListRepos`
3) `handleExternalServiceSync` then calls for an actual service sync [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/repo-updater/repoupdater/server.go?L264&subtree=true) which actually does a proper sync (with no context cancels or anything of the sort)


So the root cause is us actually calling to list all the repos when we are trying to validate the new code host config after they changed it, and then we exit prematurely after getting 1 result, causing that big dump of context cancels for all of the inflight requests.

As i mentioned above its harmless because the codehost still will sync correctly immediately after all of the context cancel errors as part of the next steps of `handleExternalServiceSync` .
Regardless we probably want to do something a bit smarter when validating our configs rather than just trying to list all of the repos lol.
